tools/pubsys/src/aws/validate_ami/mod.rs

```diff
@@ -0,0 +1,799 @@
+    let mut ami_clients = HashMap::with_capacity(expected_images.len());
+
+    for region in expected_images.keys() {
+        let client_config = build_client_config(region, base_region, &aws).await;
+        let ami_client = AmiClient::new(&client_config);
+        ami_clients.insert(region.clone(), ami_client);
+    }
+
+    // Create a map of image_id to a bool indicating whether or not the image is public. This map
+    // is needed to determine if launch permissions should be retrieved for the image (which is the
+    // case if it is not public)
+    let expected_image_public = expected_images
+        .iter()
+        .flat_map(|(_, expected_images)| {
+            expected_images
+                .iter()
+                .map(move |image_def| (image_def.id.clone(), image_def.public))
+        })
+        .collect::<HashMap<String, bool>>();
+
+    // Retrieve the EC2 images using the AmiClients
+    info!("Retrieving EC2 images");
+    let images = describe_images(
+        &ami_clients,
+        &expected_images
+            .iter()
+            .map(|(region, images)| {
+                (
+                    region.clone(),
+                    images.iter().map(|i| i.id.clone()).collect::<Vec<String>>(),
+                )
+            })
+            .collect::<HashMap<Region, Vec<String>>>(),
+        &expected_image_public,
+    )
+    .await;
+
+    // Validate the retrieved EC2 images per region
+    info!("Validating EC2 images");
+    let results: HashMap<Region, ami::Result<HashSet<AmiValidationResult>>> = images
+        .into_iter()
+        .map(|(region, region_result)| {
+            (
+                region.clone(),
+                region_result.map(|result| {
+                    validate_images_in_region(
+                        expected_images
+                            .get(region)
+                            .map(|e| e.to_owned())
+                            .unwrap_or(vec![]),
+                        &result,
+                        region,
+                    )
+                }),
+            )
+        })
+        .collect::<HashMap<Region, ami::Result<HashSet<AmiValidationResult>>>>();
+
+    let validation_results = AmiValidationResults::new(results);
+
+    // If a path was given to write the results to, write the results
+    if let Some(write_results_path) = &validate_ami_args.write_results_path {
+        // Filter the results by given status, and if no statuses were given, get all results
+        info!("Writing results to file");
+        let filtered_results = validation_results.get_results_for_status(
+            validate_ami_args
+                .write_results_filter
+                .as_ref()
+                .unwrap_or(&vec![
+                    AmiValidationResultStatus::Correct,
+                    AmiValidationResultStatus::Incorrect,
+                    AmiValidationResultStatus::Missing,
+                ]),
+        );
```

Comment:
```suggestion
        let results = if let Some(filter) = validate_ami_args.write_results_filter {
            validation_results.get_results_for_status(filter)
        else {
            validation_results.all_results()
       };
```
It's a bit unusual to list all variants of an enum as a default. You could also have `get_results_for_status` take an Option.

```suggestion
        let results = validation_results.get_results_for_status(validate_ami_args.write_results_filter);
```

Reply:
This works the same way in the ssm-validation, so I will address those together in a different PR.